### ISSUE TITLE
'Invalid magic value ...' when trying the test code

### ISSUE THREAD

**Issue by Vild** [zrho/Hydrogen#3]
Hey. I'm interested in trying to use Hydrogen for my kernel.
But now when I try to run the test code, it fails to verify the magic number.
It seems like it has mapped the memory wrong or something because it does not find the header, at the place where is should be according to the ELF64 header.

On the image below I outputted some extra information.
- The invalid magic number it found.
- The memory address where it's trying to get the magic number from.
- What the magic number should be.
- The kernel location in memory, that you get from grub.
- And lastly all the parameters for all the page_map calls.

![shot-1441837283](https://cloud.githubusercontent.com/assets/484438/9775736/e88882fe-5751-11e5-91e2-ddb52fb8bf98.png)

I can't find out what the problem could be. (I'm not that good with paging)
Any idea what could be the problem?

I've hosted the output files @ http://wild.tk/hydrogen-build.zip

**Comment by zrho** [comment]
Hey, I'm glad you consider using Hydrogen for your project :) It has been quite some time, so I have to refresh my kernel-fu to find out what's wrong; give me a couple of days, then I might be able to help you.

**Comment by vercas** [comment]
This is interesting, because the kernel blob _does_ contain the Hydrogen header and the correct values.  

Did you modify anything in the bootloader?  

Also, I can help out as well because I'm using a Hydrogen-based bootloader actively and I've messed with 80% of it.

**Comment by Vild** [comment]
I have changed nothing except adding the debug printing.
I could later try a cross compiler, because I just ran make which uses my native gcc.

**Comment by vercas** [comment]
I used 64-bit Ubuntu's GCC at first and it worked fine. (although I use one now, built under Cygwin)  
Also, what did you use to test the setup? (VM or emulation software)

**Comment by Vild** [comment]
I can give you guys my gcc version and other programs versions in a couple of hours.

But I used VirtualBox to try it.

**Comment by vercas** [comment]
I highly recommend you set VirtualBox on fire and let it burn to a crisp.  
The only thing it's good for is testing on defective hardware, because it emulates hardware bugs perfectly.  

Also, how much RAM did you give to your VM? This is very important, because too little RAM may force GRUB or Hydrogen to overwrite the _kernel64_ module.

**Comment by Vild** [comment]
I think I had 512mb, not at all sure.

**Comment by vercas** [comment]
Ah, that should be more than enough...  
I've been able to reproduce the issue in VMWare.  
Try [my GRUB build](https://github.com/vercas/Beelzebub/blob/master/image/grub2/stage2_eltorito), because it's GRUB2, which places its structures at the end of available memory, instead of the start.  

**Edit**: I'm unable to change the GRUB blob in the ISO from my current computer to test.

**Comment by Vild** [comment]
gcc 5.2.0
Virtualbox 5.0.2
QEMU failes like this when I try to run the iso. ![shot-1441909772](https://cloud.githubusercontent.com/assets/484438/9797300/af00e61c-57fa-11e5-8ab4-da8d3e75c50f.png)
But not when I run `make qemu` so I guess here that I did something wrong.
Bochs gives the same error as Virtualbox

I tried to use grub-mkrescue instead of the current iso building command, but it fails the same way.

It's not based on the ram about, I get the same error with 8GB of ram.
